compiler/src/dotty/tools/dotc/semanticdb/ExtractSemanticDB.scala

```diff
@@ -46,14 +60,141 @@ class ExtractSemanticDB extends Phase:
   // Check not needed since it does not transform trees
   override def isCheckable: Boolean = false
 
-  override def run(using Context): Unit =
-    val unit = ctx.compilationUnit
-    val extractor = Extractor()
-    extractor.extract(unit.tpdTree)
-    ExtractSemanticDB.write(unit.source, extractor.occurrences.toList, extractor.symbolInfos.toList, extractor.synthetics.toList)
+  override def runOn(units: List[CompilationUnit])(using ctx: Context): List[CompilationUnit] = {
+    val sourceRoot = ctx.settings.sourceroot.value
+    val appendDiagnostics = phaseMode == ExtractSemanticDB.PhaseMode.AppendDiagnostics
+    if (appendDiagnostics)
+      val warnings = ctx.reporter.allWarnings.groupBy(w => w.pos.source)
+      units.map { unit =>
+        val unitCtx = ctx.fresh.setCompilationUnit(unit).withRootImports
+        val outputDir =
+          ExtractSemanticDB.semanticdbPath(
+            unit.source,
+            ExtractSemanticDB.outputDirectory(using unitCtx),
+            sourceRoot
+          )
+        val source = unit.source
+        (outputDir, source)
+      }.asJava.parallelStream().forEach { case (out, source) =>
+        warnings.get(source).foreach { ws =>
+          ExtractSemanticDB.appendDiagnostics(ws.map(_.toSemanticDiagnostic), out)
```

Comment:
I think it is non deterministic to call toSemanticDiagnostic from another thread - there's lots of mutability involved.

Reply:
What do you mean? I thought the only issue was `asciiColorCodes` which should be safe.

Reply:
look at the message field in Message

Reply:
So doing something like:
```
  def messageSafe = inMessageContext(disambiguate)(msg + msgPostscript)
  /** The message to report. <nonsensical> tags are filtered out */
  @threadUnsafe lazy val message: String = messageSafe
```
would it make sense? And using `messageSafe`

Or would we need to make that whole class thread safe? Would using AtomicBoolean for the vars make any sense? How is that normally dealt with in the compiler?

Reply:
normally there is zero thread safety

Reply:
Why do we need it then? We use multiple threads for the samenticdb phase only?

Reply:
I guess it couldn't hurt to compute the message twice given that it's a separate thread? but in general the entire compiler and context etc relies on mutable fields so isn't that unsafe to read?

Reply:
Ach, I forgot we changed things here to be parallel. So should we drop that idea or maybe do `toSemanticDiagnostic` before making it parallel.